### PR TITLE
[Distributed] Enable distributed EdgeDataLoader

### PR DESCRIPTION
**Pull request by zheng-da**
## Description
This PR makes `EdgeDataLoader` work on a distributed graph. In addition, it enables `find_edges` on distributed heterogeneous graphs and `edge_subgraph` on distributed heterogeneous graphs.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/distributed/dist_graph.py [hunk 1]
```diff
@@ -976,6 +979,18 @@ def find_edges(self, edges):
         ----------
         edges : tensor
             The edge ID array.
+        eid : Int Tensor
```
Comment (classicsong) [inline python/dgl/distributed/dist_graph.py:982]: There is no eid in the param list of find_edges.

File: python/dgl/distributed/dist_graph.py [hunk 2]
```diff
@@ -984,8 +999,68 @@ def find_edges(self, edges):
...
+            num_nodes = {ntype: self.number_of_nodes(ntype) for ntype in self.ntypes}
+            subg = dgl_heterograph(subg, num_nodes_dict=num_nodes)
+        else:
+            assert len(self.etypes) == 1
+            subg = self.find_edges(edges)
+            subg = dgl_graph(subg, num_nodes=self.number_of_nodes())
+
+        if relabel_nodes:
+            subg = compact_graphs(subg)
+        assert store_ids, 'edge_subgraph always stores original node/edge IDs.'
```
Comment (classicsong) [inline python/dgl/distributed/dist_graph.py:1062]: In which function we store the original edge IDs in subg?

File: tests/distributed/test_mp_dataloader.py [hunk 3]
```diff
@@ -215,12 +215,61 @@ def start_node_dataloader(rank, tmpdir, num_server, num_workers, orig_nid, orig_
...
+        dataloader = dgl.dataloading.EdgeDataLoader(
+            dist_graph,
+            train_eid,
+            sampler,
+            batch_size=batch_size,
+            shuffle=True,
+            drop_last=False,
+            num_workers=num_workers)
+
+        groundtruth_g = CitationGraphDataset("cora")[0]
```
Comment (classicsong) [inline tests/distributed/test_mp_dataloader.py:249]: Can we add another test that uses the heterogenous dataset?

File: python/dgl/distributed/dist_graph.py (hunk 2 above)
Comment (zheng-da) [reply python/dgl/distributed/dist_graph.py:1062]: `compact_graphs` does it. i suppose `store_ids` is only valid when `relabel_nodes` is enabled?

**Review by classicsong — APPROVED** [review]
Overall LGTM.

**Pull request merged**